pkg/fleet-manager/pipeline/controller.go

```diff
@@ -0,0 +1,355 @@
+	log := ctrl.LoggerFrom(ctx)
+
+	// Render the trigger.
+	triggerResource, err := render.RenderTriggerWithPipeline(pipeline)
+	if err != nil {
+		log.Error(err, "Error rendering the trigger")
+		return ctrl.Result{}, err
+	}
+
+	// Apply the trigger resources.
+	if _, err := util.PatchResources(triggerResource); err != nil {
+		log.Error(err, "Error applying trigger resources")
+		return ctrl.Result{}, errors.Wrapf(err, "failed to apply trigger resources")
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// reconcilePipelineStatus updates the status of the pipeline.
+func (p *PipelineManager) reconcilePipelineStatus(ctx context.Context, pipeline *pipelineapi.Pipeline) (ctrl.Result, error) {
+	// Update event listener service name in the status.
+	pipeline.Status.EventListenerServiceName = getListenerServiceName(pipeline)
+	return ctrl.Result{}, nil
+}
+
+// reconcileDeletePipeline handles the deletion of a Pipeline object.
+func (p *PipelineManager) reconcileDeletePipeline(ctx context.Context, pipeline *pipelineapi.Pipeline) (ctrl.Result, error) {
+	// First, delete all Pods with the specific label.
+	if err := p.deleteAssociatedPods(ctx, pipeline.Namespace, pipeline.Name); err != nil {
+		return ctrl.Result{}, fmt.Errorf("error deleting associated pods: %v", err)
+	}
+
+	// After successful deletion of Pods, remove the finalizer from the Pipeline.
+	controllerutil.RemoveFinalizer(pipeline, PipelineFinalizer)
+
+	return ctrl.Result{}, nil
+}
+
+// deleteAssociatedPods deletes all Pods in the same namespace as the Pipeline
+// that have a label with key TektonPipelineLabel and value equal to pipelineName.
+func (p *PipelineManager) deleteAssociatedPods(ctx context.Context, namespace, pipelineName string) error {
+	// The Kurator pipeline name is equal to the Tekton pipeline name, so we can use MatchingLabels{TektonPipelineLabel: pipelineName}
+	labelSelector := client.MatchingLabels{TektonPipelineLabel: pipelineName}
+	var pods corev1.PodList
+
+	// List all Pods in the same namespace with the specified label selector.
+	if err := p.Client.List(ctx, &pods, client.InNamespace(namespace), labelSelector); err != nil {
+		return fmt.Errorf("error listing pods: %v", err)
+	}
+
+	// Delete each found Pod.
+	for _, pod := range pods.Items {
+		// Delete the Pod using Foreground deletion policy.
+		// This ensures that all dependent resources like PVCs are also deleted before the Pod itself is deleted.
+		err := p.Client.Delete(ctx, &pod, client.PropagationPolicy(metav1.DeletePropagationForeground))
+		if err != nil && !apiserrors.IsNotFound(err) {
+			// If the error is not a NotFound error, return the error.
+			return fmt.Errorf("error deleting pod %s: %v", pod.Name, err)
+		}
+	}
+
+	return nil
+}
+
+// isRBACResourceReady checks if necessary RBAC resources are ready.
+func (p *PipelineManager) isRBACResourceReady(ctx context.Context, rbacConfig render.RBACConfig) bool {
```

Comment:
donot you need to check clusterRole and Role?

Reply:
I am not sure I got it,
do you mean checking the content of ServiceAccount?

Reply:
the existence of them

Reply:
clusterRole and Role is provider by Tekton, so we do not need check them in Kurator controller.